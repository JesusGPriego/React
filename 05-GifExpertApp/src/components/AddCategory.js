import React from "react";
import { useState } from "react";
//proptypes
import PropTypes from "prop-types";

//getting the funciton (allowing component comunication)
/**
 * Add category allow the user to add more categories to the category list.
 * it is a simple input text, that add another category to the listo when 'enter' is pushed.
 * @param {setCategories} is the function that is triggered whenever enter is pushed. Another category is added so, another grid item will be created.
 */
export const AddCategory = ({ setCategories }) => {
  const [inputValue, setinputValue] = useState(""); //not leave empty, or undefined is set.

  const handleInputChange = (e) => {
    // console.log(e.target.value);
    //input value is all the text written, so by setting this value to input value, we actually got the input value.
    setinputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    // console.log("Submitted");
    if (inputValue.trim().length > 2) {
      setCategories((cats) => [inputValue, ...cats]);
      setinputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChange}
        placeholder="Insert new category"
      />
    </form>
  );
};

AddCategory.propTypes = {
  //setting the function as required:
  setCategories: PropTypes.func.isRequired,
};
