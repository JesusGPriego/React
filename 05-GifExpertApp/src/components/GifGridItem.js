import React from "react";
import PropTypes from "prop-types";
/**
 * Creates a grid item for the GifGrid. 
 * This item display an image and its title.
 * 
 * @param {id} is the image id.
 * @param {title} it is the image title.
 * @param {url} it is the image's url.
 * @returns a card that contains an image, and the image's title.
 */
export const GifGridItem = ({ title, url }) => {
  return (
    <figure className="card animate__animated animate__fadeIn">
      <img src={url} alt={title} />
      <figcaption>{title}</figcaption>
    </figure>
  );
};


GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
}