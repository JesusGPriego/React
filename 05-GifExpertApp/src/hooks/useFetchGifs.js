import { useState, useEffect } from "react"
import {getGifs} from '../helpers/getGifs'

/**
 * Custom hook that calls 'getGifs' and serve info to the component that calls it.
 * In this case, this hook is called by GigGrid to create 10 GifGridItem with its info.
 * @param {*} category the topic that the user is looking for.
 * @returns an object with imgs data and a boolean 'loading' which allow the dev to dinamically
 * change the web appearence.
 */
export const useFetchGifs = (category) =>{
    const [state, setState] = useState(
        {
            data:[],
            loading:true
        }
    );

    //use effect allow us to execute the code just one time, by setting second argument as an empty array.
    useEffect(() => {
        getGifs(category)
            .then(imgs => {
                setState({
                    data: imgs,
                    loading:false,
                });
            });
    }, []);


    // setTimeout(() =>{
    //     setState({
    //         data: [1,2,3,4,5],
    //         loading:false,
    //     })
    // }, 1500)

    return state; //state is data and loading.
}
