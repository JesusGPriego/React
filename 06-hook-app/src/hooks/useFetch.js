import {useState, useEffect, useRef} from 'react';

export const useFetch = (url) => {
    //when the component is created, the component is mounted 100%
    const isMounted = useRef(true);
    const [state, setState] = useState({data:null, loading:true, error:null});

    //in case the component is unmounted, the var is mounted turns into false:
    useEffect(() => {
        return () =>{
            isMounted.current = false;
        };
    }, []);

    useEffect(() => {
        setState({data:null, loading:true, error:null});
        fetch(url)
            .then(response => response.json())
            .then(data =>{

                //if the component is not mounted, setState won't trigger. 
                if(isMounted.current){
                    setState({
                        loading:false,
                        error:null,
                        data:data,
                    });
                }
                
            });

    },[url]);

    return state;

};
