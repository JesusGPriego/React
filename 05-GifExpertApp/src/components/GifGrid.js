import React from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
/**
 * Creates a grid of images that match the topic that the user is looking for.
 * Topic here is the category.
 * 
 * @param {category} is what the user is looking for.  
 * @returns a new fragment, which is a grid with 10 images.
 */
export const GifGrid = ({ category }) => {
  //asking custom hook for info:
  const {data:images, loading} = useFetchGifs(category);

  return (
    <>
    {loading && <p>Cargando...</p>}
      <h3>{category}</h3>
      <div className="card-grid animate__animated animate__fadeIn">
        {images.map((image) => (
          <GifGridItem key={image.id} {...image} />
        ))}
      </div>
    </>
  );
};
