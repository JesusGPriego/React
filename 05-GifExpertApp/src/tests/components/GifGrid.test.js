import { shallow } from "enzyme"
import {GifGrid} from '../../components/GifGrid';
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock('../../hooks/useFetchGifs');

describe('Testing GifGrid Component', () => {


    const category = 'One Punch Man'
    
    test('should return a GifGrid that match the snapshot.', () => {  

        useFetchGifs.mockReturnValue({
            data:[],
            loading:true
        })
        const wrapper = shallow(<GifGrid category={category}/>)
        expect(wrapper).toMatchSnapshot()

    })
    

    test('should show item when images are loaded with "useFetchGifs"', () => {
        
        const gifs = [
            {
                id:'abc',
                url:'https://myimage.jpg',
                title: 'whatever title',
            }
        ]
        //faking data loaded:
        useFetchGifs.mockReturnValue({
            data:gifs,
            loading:false,
        })
        const wrapper = shallow(<GifGrid category={category}/>)
        expect(wrapper).toMatchSnapshot();
        //no loading p should exist after data is loaded:
        expect(wrapper.find('p').exists()).toBe(false);
        //asserting that there is the same gifgriditems as gifs item:
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })
    

})
