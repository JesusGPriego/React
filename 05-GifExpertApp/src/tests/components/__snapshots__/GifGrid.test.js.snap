// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifGrid Component should return a GifGrid that match the snapshot. 1`] = `
<Fragment>
  <p>
    Cargando...
  </p>
  <h3>
    One Punch Man
  </h3>
  <div
    className="card-grid animate__animated animate__fadeIn"
  />
</Fragment>
`;

exports[`Testing GifGrid Component should show item when images are loaded with "useFetchGifs" 1`] = `
<Fragment>
  <h3>
    One Punch Man
  </h3>
  <div
    className="card-grid animate__animated animate__fadeIn"
  >
    <GifGridItem
      id="abc"
      key="abc"
      title="whatever title"
      url="https://myimage.jpg"
    />
  </div>
</Fragment>
`;
